### ISSUE TITLE
Vertical space is wrong when \multicolumn spans the last column

### ISSUE THREAD

**Issue by eg9** [latex3/latex2e#42]
## Brief outline of the bug

When a `\multicolumn` spans the last column of a `tabular`, the vertical space in ```\\[<skip>]``` is not honored.

## Minimal example showing the bug

```latex
\RequirePackage{latexbug}
\documentclass{article}
\usepackage{array}
\begin{document}
\begin{tabular}{lll}
  row 1 & \multicolumn{2}{l}{multi} \\[\medskipamount]
  row 2 & a & b                     \\
  row 3 & c & d                     \\[\medskipamount]
  row 4 & a & b                     \\
  row 5 & a & b                     \\
\end{tabular}
\end{document}
```

## Log file (required) and possibly PDF file  
```
This is pdfTeX, Version 3.14159265-2.6-1.40.19 (TeX Live 2018) (preloaded format=pdflatex 2018.4.30)  30 APR 2018 12:44
entering extended mode
 restricted \write18 enabled.
 %&-line parsing enabled.
**ufarr
(./ufarr.tex
LaTeX2e <2018-04-01> patch level 3
Babel <3.19> and hyphenation patterns for 84 language(s) loaded.
(/usr/local/texlive/2018/texmf-dist/tex/latex/latexbug/latexbug.sty
Package: latexbug 2017/10/12 v1.0d Bug-classification
)
(/usr/local/texlive/2018/texmf-dist/tex/latex/base/article.cls
Document Class: article 2014/09/29 v1.4h Standard LaTeX document class
(/usr/local/texlive/2018/texmf-dist/tex/latex/base/size10.clo
File: size10.clo 2014/09/29 v1.4h Standard LaTeX file (size option)
)
\c@part=\count80
\c@section=\count81
\c@subsection=\count82
\c@subsubsection=\count83
\c@paragraph=\count84
\c@subparagraph=\count85
\c@figure=\count86
\c@table=\count87
\abovecaptionskip=\skip41
\belowcaptionskip=\skip42
\bibindent=\dimen102
)
(/usr/local/texlive/2018/texmf-dist/tex/latex/tools/array.sty
Package: array 2018/04/07 v2.4g Tabular extension package (FMi)
\col@sep=\dimen103
\ar@mcellbox=\box26
\extrarowheight=\dimen104
\NC@list=\toks14
\extratabsurround=\skip43
\backup@length=\skip44
\ar@cellbox=\box27
) (./ufarr.aux)
\openout1 = `ufarr.aux'.

LaTeX Font Info:    Checking defaults for OML/cmm/m/it on input line 4.
LaTeX Font Info:    ... okay on input line 4.
LaTeX Font Info:    Checking defaults for T1/cmr/m/n on input line 4.
LaTeX Font Info:    ... okay on input line 4.
LaTeX Font Info:    Checking defaults for OT1/cmr/m/n on input line 4.
LaTeX Font Info:    ... okay on input line 4.
LaTeX Font Info:    Checking defaults for OMS/cmsy/m/n on input line 4.
LaTeX Font Info:    ... okay on input line 4.
LaTeX Font Info:    Checking defaults for OMX/cmex/m/n on input line 4.
LaTeX Font Info:    ... okay on input line 4.
LaTeX Font Info:    Checking defaults for U/cmr/m/n on input line 4.
LaTeX Font Info:    ... okay on input line 4.
LaTeX Font Info:    External font `cmex10' loaded for size
(Font)              <7> on input line 5.
LaTeX Font Info:    External font `cmex10' loaded for size
(Font)              <5> on input line 5.

[1

{/usr/local/texlive/2018/texmf-var/fonts/map/pdftex/updmap/pdftex.map}]
(./ufarr.aux) ) 
Here is how much of TeX's memory you used:
 757 strings out of 492649
 12936 string characters out of 6129622
 65839 words of memory out of 5000000
 4718 multiletter control sequences out of 15000+600000
 3640 words of font info for 14 fonts, out of 8000000 for 9000
 1141 hyphenation exceptions out of 8191
 23i,7n,17p,139b,486s stack positions out of 5000i,500n,10000p,200000b,80000s
</usr/local/texlive/2018/texmf-dist/fonts/type1/public/amsfonts/
cm/cmr10.pfb>
Output written on ufarr.pdf (1 page, 13658 bytes).
PDF statistics:
 12 PDF objects out of 1000 (max. 8388607)
 7 compressed objects within 1 object stream
 0 named destinations out of 1000 (max. 500000)
 1 words of extra memory for PDF output out of 10000 (max. 10000000)
```

## PDF file
[ufarr.pdf](https://github.com/latex3/latex2e/files/1960257/ufarr.pdf)

## PNG image
<img width="154" alt="screen shot 2018-04-30 at 12 51 18" src="https://user-images.githubusercontent.com/302839/39424080-3ae85554-4c75-11e8-86f1-f1919f2f32ca.png">

## PNG Image obtained with `array` version 2.4d (on TeX Live 2017)
<img width="148" alt="screen shot 2018-04-30 at 12 52 14" src="https://user-images.githubusercontent.com/302839/39424145-8062b12e-4c75-11e8-96eb-f0d78bffb113.png">

**Comment by u-fischer** [comment]
related problems with vertical spaces (the second one can also be seen with tabularx instead of tabu):

https://tex.stackexchange.com/q/429272/87678
https://tex.stackexchange.com/questions/429307/problem-with-new-array-sty-colortbl-sty-and-tabu-sty

**Comment by FrankMittelbach** [comment]
Fix for the issues is most likely
```
\def\@xargarraycr#1{\unskip\gdef\do@row@strut
 {\@tempdima #1\advance\@tempdima \dp\@arstrutbox
   \vrule \@depth\@tempdima \@width\z@\global\let\do@row@strut\relax}%
   \ifnum\@multicnt >\z@ \do@row@strut \fi\cr}
```
though that needs a bit more testing.

The colortbl patch as suggested by David in a rush may not be quite right but once done probably also fixes the other two mentioned cases (though not 100% sure of the tabu case yet)

**Comment by FrankMittelbach** [comment]
In addition to what @davidcarlisle  already suggested as a patch to colortbl it needs a few more lines:
```
\def\@classz{\@classx
   \@tempcnta \count@
   \prepnext@tok
\expandafter\CT@extract\the\toks\@tempcnta\columncolor!\@nil
   \@addtopreamble{%
    \setbox\z@\hbox\bgroup\bgroup
      \CT@everycr{}%
      \ifcase \@chnum
      \hskip\stretch{.5}\kern\z@
      \d@llarbegin
      \insert@column
      \d@llarend\do@row@strut\hskip\stretch{.5}\or
      \d@llarbegin \insert@column \d@llarend\do@row@strut \hfill \or
      \hfill\kern\z@ \d@llarbegin \insert@column \d@llarend\do@row@strut \or
%% FMi: replace      
\setbox\ar@mcellbox\vbox
       \@startpbox{\@nextchar}\insert@column \@endpbox
    \ar@align@mcell
    \do@row@strut \or
%    $\vcenter
%   \@startpbox{\@nextchar}\insert@column \@endpbox $\or
    \vtop \@startpbox{\@nextchar}\insert@column \@endpbox
%% FMi add     
    \do@row@strut
    \or
    \vbox \@startpbox{\@nextchar}\insert@column \@endpbox
%% FMi add      
    \do@row@strut
  \fi
 \egroup\egroup
\begingroup
  \CT@setup
  \CT@column@color
  \CT@row@color
  \CT@cell@color
  \CT@do@color
\endgroup
        \@tempdima\ht\z@
        \advance\@tempdima\minrowclearance
        \vrule\@height\@tempdima\@width\z@
        \unhbox\z@}%
  \prepnext@tok}
```
after that the examples on SX run for me.

**Comment by dcpurton** [comment]
I still get slightly different output for this `tabu` example:

```TeX
\documentclass{article}
\usepackage{colortbl}
\usepackage{tabu}
\pagestyle{empty}
\begin{document}
\begin{tabu}{|l|X|}
  \hline
  Row 1 & \\
  \hline
  Row 2 & \\[5mm]
  \hline
  Row 3 & \\[5mm]
  \hline
  Row 4 & \\[5mm]
  \hline
\end{tabu}
\end{document}
```

With `array  2016/10/06 v2.4d` and `colortbl 2012/02/13 v1.0a` I get:

![old](https://user-images.githubusercontent.com/6268044/39458398-56c35896-4d33-11e8-8b9d-fb4bc48e3c44.png)

With `array 2018/04/07 v2.4g` including the patch in this thread and `colortbl` from https://tex.stackexchange.com/q/429272/87678 with the patch from this thread added I get:

![new](https://user-images.githubusercontent.com/6268044/39458470-f47c507e-4d33-11e8-9bd4-2ef929ac82c1.png)

**Comment by u-fischer** [comment]
This looks as if the new code corrects a bug in tabu. In your first output  tabu added 10mm instead of 5mm.

**Comment by FrankMittelbach** [comment]
@dcpurton I think @u-fischer is right: looking into the generated page one sees a rule of with a depth of roughly twice of what is requested (plus roughly a baselineskip). With the new version (including the patches) that is gone. Replacing tabu with tabularx doesn't show that behavior (ie fine in both versions).

**Comment by dcpurton** [comment]
@u-fischer and @FrankMittelbach, I agree, the old behaviour of `tabu` does seem wrong (and I've even wondered about it before). It's a pity it will break old documents, but maybe that can't be helped. Thanks for dealing with this.

**Comment by FrankMittelbach** [comment]
I just fixed the `array` part of the problem and will upload that the moment the internet connection is a bit more stable over here. I also arranged for the package to rollback to the 2017 behavior if there is a rollback request so that older documents should still show the wrong (but probably used) behavior with `tabu` and the like.
In addition colortbl needs changing (as outlined above) so I pass that on to @davidcarlisle.
